{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycirclize import Circos\n",
    "from pycirclize.utils import ColorCycler, load_eukaryote_example_dataset\n",
    "import numpy as np\n",
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Gene Amplifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr\n",
      "chr1     8\n",
      "chr19    7\n",
      "chr12    7\n",
      "chr8     6\n",
      "chr2     5\n",
      "chr17    4\n",
      "chr3     3\n",
      "chr20    3\n",
      "chr6     3\n",
      "chr18    3\n",
      "chr14    2\n",
      "chr11    2\n",
      "chr10    2\n",
      "chr7     2\n",
      "chr5     2\n",
      "chr4     2\n",
      "chr15    1\n",
      "chr22    1\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>q values</th>\n",
       "      <th>Description</th>\n",
       "      <th>negLogQ</th>\n",
       "      <th>Chromosome_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39109153</td>\n",
       "      <td>41523062</td>\n",
       "      <td>1.280000e-21</td>\n",
       "      <td>1p34.2(MYCL1)</td>\n",
       "      <td>20.892790</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148646436</td>\n",
       "      <td>149486905</td>\n",
       "      <td>5.190000e-18</td>\n",
       "      <td>1q21.2(MCL1)</td>\n",
       "      <td>17.284833</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>232526850</td>\n",
       "      <td>233617929</td>\n",
       "      <td>8.670000e-07</td>\n",
       "      <td>1q42.3(IRF2BP2)</td>\n",
       "      <td>6.061981</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154088696</td>\n",
       "      <td>154329131</td>\n",
       "      <td>3.880000e-09</td>\n",
       "      <td>---</td>\n",
       "      <td>8.411168</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36338557</td>\n",
       "      <td>36480225</td>\n",
       "      <td>2.120000e-11</td>\n",
       "      <td>1p34.3</td>\n",
       "      <td>10.673664</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr19</td>\n",
       "      <td>20823633</td>\n",
       "      <td>21917935</td>\n",
       "      <td>3.310000e-05</td>\n",
       "      <td>---</td>\n",
       "      <td>4.480172</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr20</td>\n",
       "      <td>45338887</td>\n",
       "      <td>45497265</td>\n",
       "      <td>4.886900e-04</td>\n",
       "      <td>ZMYND8</td>\n",
       "      <td>3.310967</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr20</td>\n",
       "      <td>61411961</td>\n",
       "      <td>62435964</td>\n",
       "      <td>9.720000e-06</td>\n",
       "      <td>20q13.33</td>\n",
       "      <td>5.012334</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr20</td>\n",
       "      <td>29703246</td>\n",
       "      <td>29808748</td>\n",
       "      <td>3.474400e-04</td>\n",
       "      <td>BCL2L1</td>\n",
       "      <td>3.459120</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chr22</td>\n",
       "      <td>28450357</td>\n",
       "      <td>28767355</td>\n",
       "      <td>2.076100e-04</td>\n",
       "      <td>MTMR3</td>\n",
       "      <td>3.682752</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr      Start        End      q values      Description    negLogQ  \\\n",
       "0    chr1   39109153   41523062  1.280000e-21    1p34.2(MYCL1)  20.892790   \n",
       "1    chr1  148646436  149486905  5.190000e-18     1q21.2(MCL1)  17.284833   \n",
       "2    chr1  232526850  233617929  8.670000e-07  1q42.3(IRF2BP2)   6.061981   \n",
       "3    chr1  154088696  154329131  3.880000e-09              ---   8.411168   \n",
       "4    chr1   36338557   36480225  2.120000e-11           1p34.3  10.673664   \n",
       "..    ...        ...        ...           ...              ...        ...   \n",
       "58  chr19   20823633   21917935  3.310000e-05              ---   4.480172   \n",
       "61  chr20   45338887   45497265  4.886900e-04           ZMYND8   3.310967   \n",
       "59  chr20   61411961   62435964  9.720000e-06         20q13.33   5.012334   \n",
       "60  chr20   29703246   29808748  3.474400e-04           BCL2L1   3.459120   \n",
       "62  chr22   28450357   28767355  2.076100e-04            MTMR3   3.682752   \n",
       "\n",
       "    Chromosome_num  \n",
       "0                1  \n",
       "1                1  \n",
       "2                1  \n",
       "3                1  \n",
       "4                1  \n",
       "..             ...  \n",
       "58              19  \n",
       "61              20  \n",
       "59              20  \n",
       "60              20  \n",
       "62              22  \n",
       "\n",
       "[63 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = './amp.csv'\n",
    "\n",
    "df = pd.read_csv(file_path, usecols=['Chromosome', 'Wide peak start', 'Wide peak end', 'q values', 'Description'])\n",
    "\n",
    "df[\"negLogQ\"] = np.log10(df[\"q values\"]) * -1\n",
    "\n",
    "def extract_numeric_part(chromosome):\n",
    "    return int(chromosome[3:])\n",
    "\n",
    "df['Chromosome_num'] = df['Chromosome'].apply(extract_numeric_part)\n",
    "\n",
    "# Sort the DataFrame based on the numeric part of the chromosome names\n",
    "df = df.sort_values(by='Chromosome_num')\n",
    "\n",
    "df.rename(columns={'Chromosome': 'Chr', \n",
    "                       'Wide peak start': 'Start', \n",
    "                       'Wide peak end': 'End'}, inplace=True)\n",
    "\n",
    "\n",
    "print(df[\"Chr\"].value_counts())\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "At least one row with chromosome chr1 is present.\n",
      "At least one row with chromosome chr2 is present.\n",
      "At least one row with chromosome chr3 is present.\n",
      "At least one row with chromosome chr4 is present.\n",
      "At least one row with chromosome chr5 is present.\n",
      "At least one row with chromosome chr6 is present.\n",
      "At least one row with chromosome chr7 is present.\n",
      "At least one row with chromosome chr8 is present.\n",
      "No row with chromosome chr9 is present.\n",
      "At least one row with chromosome chr10 is present.\n",
      "At least one row with chromosome chr11 is present.\n",
      "At least one row with chromosome chr12 is present.\n",
      "No row with chromosome chr13 is present.\n",
      "At least one row with chromosome chr14 is present.\n",
      "At least one row with chromosome chr15 is present.\n",
      "No row with chromosome chr16 is present.\n",
      "At least one row with chromosome chr17 is present.\n",
      "At least one row with chromosome chr18 is present.\n",
      "At least one row with chromosome chr19 is present.\n",
      "At least one row with chromosome chr20 is present.\n",
      "No row with chromosome chr21 is present.\n",
      "At least one row with chromosome chr22 is present.\n",
      "No row with chromosome chrX is present.\n",
      "No row with chromosome chrY is present.\n"
     ]
    }
   ],
   "source": [
    "def check_chromosome_presence(df):\n",
    "    # List of chromosome names to check\n",
    "    chromosomes_to_check = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', \n",
    "                            'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', \n",
    "                            'chr20', 'chr21', 'chr22', 'chrX', 'chrY']\n",
    "\n",
    "    # Check if any of the chromosome names are present in the DataFrame\n",
    "    for chromosome in chromosomes_to_check:\n",
    "        if chromosome in df['Chr'].values:\n",
    "            print(f\"At least one row with chromosome {chromosome} is present.\")\n",
    "        else:\n",
    "            print(f\"No row with chromosome {chromosome} is present.\")\n",
    "\n",
    "check_chromosome_presence(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file_path = './amp_sorted.csv'\n",
    "df.to_csv(output_file_path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Gene Deletions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr\n",
      "chr5     5\n",
      "chr1     4\n",
      "chr19    4\n",
      "chr16    4\n",
      "chr3     3\n",
      "chr6     3\n",
      "chr17    3\n",
      "chr9     3\n",
      "chr13    2\n",
      "chr21    2\n",
      "chr15    2\n",
      "chr11    2\n",
      "chr2     2\n",
      "chr10    2\n",
      "chr8     2\n",
      "chr4     2\n",
      "chr12    1\n",
      "chr14    1\n",
      "chr18    1\n",
      "chr7     1\n",
      "chr22    1\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>q values</th>\n",
       "      <th>Description</th>\n",
       "      <th>negLogQ</th>\n",
       "      <th>Chromosome_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16666761</td>\n",
       "      <td>28781645</td>\n",
       "      <td>1.380000e-18</td>\n",
       "      <td>1p36.11</td>\n",
       "      <td>17.860121</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>23549970</td>\n",
       "      <td>7.780000e-11</td>\n",
       "      <td>---</td>\n",
       "      <td>10.109020</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56202184</td>\n",
       "      <td>118877972</td>\n",
       "      <td>1.073600e-01</td>\n",
       "      <td>---</td>\n",
       "      <td>0.969157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>204477622</td>\n",
       "      <td>220287854</td>\n",
       "      <td>1.785300e-01</td>\n",
       "      <td>---</td>\n",
       "      <td>0.748289</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>234758443</td>\n",
       "      <td>242951149</td>\n",
       "      <td>8.740000e-07</td>\n",
       "      <td>2q37.3</td>\n",
       "      <td>6.058489</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chr      Start        End      q values Description    negLogQ  \\\n",
       "0  chr1   16666761   28781645  1.380000e-18     1p36.11  17.860121   \n",
       "1  chr1          1   23549970  7.780000e-11         ---  10.109020   \n",
       "2  chr1   56202184  118877972  1.073600e-01         ---   0.969157   \n",
       "3  chr1  204477622  220287854  1.785300e-01         ---   0.748289   \n",
       "4  chr2  234758443  242951149  8.740000e-07      2q37.3   6.058489   \n",
       "\n",
       "   Chromosome_num  \n",
       "0               1  \n",
       "1               1  \n",
       "2               1  \n",
       "3               1  \n",
       "4               2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "file_path = './del.csv'\n",
    "\n",
    "del_df = pd.read_csv(file_path, usecols=['Chromosome', 'Wide peak start', 'Wide peak end', 'q values', 'Description'])\n",
    "del_df[\"negLogQ\"] = np.log10(del_df[\"q values\"]) * -1\n",
    "del_df['Chromosome_num'] = del_df['Chromosome'].apply(extract_numeric_part)\n",
    "\n",
    "del_df.rename(columns={'Chromosome': 'Chr', \n",
    "                       'Wide peak start': 'Start', \n",
    "                       'Wide peak end': 'End'}, inplace=True)\n",
    "\n",
    "# Sort the DataFrame based on the numeric part of the chromosome names\n",
    "del_df = del_df.sort_values(by='Chromosome_num')\n",
    "print(del_df[\"Chr\"].value_counts())\n",
    "del_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "At least one row with chromosome chr1 is present.\n",
      "At least one row with chromosome chr2 is present.\n",
      "At least one row with chromosome chr3 is present.\n",
      "At least one row with chromosome chr4 is present.\n",
      "At least one row with chromosome chr5 is present.\n",
      "At least one row with chromosome chr6 is present.\n",
      "At least one row with chromosome chr7 is present.\n",
      "At least one row with chromosome chr8 is present.\n",
      "At least one row with chromosome chr9 is present.\n",
      "At least one row with chromosome chr10 is present.\n",
      "At least one row with chromosome chr11 is present.\n",
      "At least one row with chromosome chr12 is present.\n",
      "At least one row with chromosome chr13 is present.\n",
      "At least one row with chromosome chr14 is present.\n",
      "At least one row with chromosome chr15 is present.\n",
      "At least one row with chromosome chr16 is present.\n",
      "At least one row with chromosome chr17 is present.\n",
      "At least one row with chromosome chr18 is present.\n",
      "At least one row with chromosome chr19 is present.\n",
      "No row with chromosome chr20 is present.\n",
      "At least one row with chromosome chr21 is present.\n",
      "At least one row with chromosome chr22 is present.\n",
      "No row with chromosome chrX is present.\n",
      "No row with chromosome chrY is present.\n"
     ]
    }
   ],
   "source": [
    "check_chromosome_presence(del_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file_path = './del_sorted.csv'\n",
    "del_df.to_csv(output_file_path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create GFF files for arrow circos plot with gene description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>q values</th>\n",
       "      <th>Description</th>\n",
       "      <th>negLogQ</th>\n",
       "      <th>Chromosome_num</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39109153</td>\n",
       "      <td>41523062</td>\n",
       "      <td>1.280000e-21</td>\n",
       "      <td>1p34.2(MYCL1)</td>\n",
       "      <td>20.892790</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148646436</td>\n",
       "      <td>149486905</td>\n",
       "      <td>5.190000e-18</td>\n",
       "      <td>1q21.2(MCL1)</td>\n",
       "      <td>17.284833</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>232526850</td>\n",
       "      <td>233617929</td>\n",
       "      <td>8.670000e-07</td>\n",
       "      <td>1q42.3(IRF2BP2)</td>\n",
       "      <td>6.061981</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154088696</td>\n",
       "      <td>154329131</td>\n",
       "      <td>3.880000e-09</td>\n",
       "      <td>---</td>\n",
       "      <td>8.411168</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36338557</td>\n",
       "      <td>36480225</td>\n",
       "      <td>2.120000e-11</td>\n",
       "      <td>1p34.3</td>\n",
       "      <td>10.673664</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>chr19</td>\n",
       "      <td>1</td>\n",
       "      <td>306931</td>\n",
       "      <td>3.030000e-56</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>55.518557</td>\n",
       "      <td>19</td>\n",
       "      <td>del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>chr19</td>\n",
       "      <td>45712960</td>\n",
       "      <td>55025072</td>\n",
       "      <td>4.274900e-02</td>\n",
       "      <td>---</td>\n",
       "      <td>1.369074</td>\n",
       "      <td>19</td>\n",
       "      <td>del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>chr21</td>\n",
       "      <td>41501909</td>\n",
       "      <td>46944323</td>\n",
       "      <td>4.164300e-02</td>\n",
       "      <td>---</td>\n",
       "      <td>1.380458</td>\n",
       "      <td>21</td>\n",
       "      <td>del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>chr21</td>\n",
       "      <td>1</td>\n",
       "      <td>23298229</td>\n",
       "      <td>1.032700e-04</td>\n",
       "      <td>---</td>\n",
       "      <td>3.986026</td>\n",
       "      <td>21</td>\n",
       "      <td>del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>chr22</td>\n",
       "      <td>49443563</td>\n",
       "      <td>49691432</td>\n",
       "      <td>1.010000e-39</td>\n",
       "      <td>22q13.33</td>\n",
       "      <td>38.995679</td>\n",
       "      <td>22</td>\n",
       "      <td>del</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chr      Start        End      q values      Description    negLogQ  \\\n",
       "0     chr1   39109153   41523062  1.280000e-21    1p34.2(MYCL1)  20.892790   \n",
       "1     chr1  148646436  149486905  5.190000e-18     1q21.2(MCL1)  17.284833   \n",
       "2     chr1  232526850  233617929  8.670000e-07  1q42.3(IRF2BP2)   6.061981   \n",
       "3     chr1  154088696  154329131  3.880000e-09              ---   8.411168   \n",
       "4     chr1   36338557   36480225  2.120000e-11           1p34.3  10.673664   \n",
       "..     ...        ...        ...           ...              ...        ...   \n",
       "108  chr19          1     306931  3.030000e-56          19p13.3  55.518557   \n",
       "109  chr19   45712960   55025072  4.274900e-02              ---   1.369074   \n",
       "110  chr21   41501909   46944323  4.164300e-02              ---   1.380458   \n",
       "111  chr21          1   23298229  1.032700e-04              ---   3.986026   \n",
       "112  chr22   49443563   49691432  1.010000e-39         22q13.33  38.995679   \n",
       "\n",
       "     Chromosome_num type  \n",
       "0                 1  amp  \n",
       "1                 1  amp  \n",
       "2                 1  amp  \n",
       "3                 1  amp  \n",
       "4                 1  amp  \n",
       "..              ...  ...  \n",
       "108              19  del  \n",
       "109              19  del  \n",
       "110              21  del  \n",
       "111              21  del  \n",
       "112              22  del  \n",
       "\n",
       "[113 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge amp and del genes\n",
    "df[\"type\"] = \"amp\"\n",
    "del_df[\"type\"] = \"del\"\n",
    "frames = [df, del_df]\n",
    "\n",
    "result = pd.concat(frames, ignore_index=True)\n",
    "result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./gene_GFF.txt'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create gff file\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Mapping the type to GFF strand symbol\n",
    "result['Strand'] = result['type'].apply(lambda x: '+' if x == 'amp' else '-')\n",
    "\n",
    "result.to_csv(\"./final_gene.csv\")\n",
    "\n",
    "# Creating the GFF format as specified\n",
    "gff_df = pd.DataFrame({\n",
    "    \"seqname\": \"ovarian\",\n",
    "    \"source\": \"python\",\n",
    "    \"feature\": result[\"type\"],\n",
    "    \"start\": result[\"Start\"],\n",
    "    \"end\": result[\"End\"],\n",
    "    \"score\": result[\"negLogQ\"],\n",
    "    \"strand\": result['Strand'],\n",
    "    \"frame\": \".\",\n",
    "    \"attribute\": \"ID=\" + result[\"Chr\"] + \";comments=\" + result[\"Chr\"] + \"_\" + result[\"Description\"]\n",
    "})\n",
    "\n",
    "# Creating the GFF formatted content\n",
    "gff_content = gff_df.to_csv(sep='\\t', index=False, header=False)\n",
    "gff_file_path = './gene_GFF.txt'\n",
    "with open(gff_file_path, 'w') as file:\n",
    "    file.write(gff_content)\n",
    "\n",
    "gff_file_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Still need to save txt file for compare circos plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./compare_pre_GFF.txt'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"final_gene.csv\")\n",
    "df.drop(columns=df.columns[0], axis=1, inplace=True)\n",
    "df.drop(columns=df.columns[4], axis=1, inplace=True)\n",
    "df.drop(columns=df.columns[5], axis=1, inplace=True)\n",
    "df.drop(columns=df.columns[6], axis=1, inplace=True)\n",
    "\n",
    "df['dist'] = df['End'] - df['Start']\n",
    "\n",
    "# Determine the maximum distance for each chromosome and add it as a new column\n",
    "df['max_distance'] = df.groupby('Chr')['dist'].transform('max')\n",
    "\n",
    "# Add the minimum start value for each chromosome to the DataFrame\n",
    "df['min_start'] = df.groupby('Chr')['Start'].transform('min')\n",
    "\n",
    "# Add the maximum end value for each chromosome to the DataFrame\n",
    "df['max_end'] = df.groupby('Chr')['End'].transform('max')\n",
    "\n",
    "df['n_start'] = ((df['Start'] - df['min_start']) * df['max_distance']) / (df['max_end'] - df['min_start'])\n",
    "df['n_end'] = ((df['End'] - df['min_start']) * df['max_distance']) / (df['max_end'] - df['min_start'])\n",
    "\n",
    "\n",
    "com_df = pd.DataFrame({\n",
    "    \"seqname\": df[\"Chr\"],\n",
    "    \"source\": \"ovarian\",\n",
    "    \"feature\": df[\"type\"],\n",
    "    \"start\": df[\"n_start\"].astype(int).clip(lower=1),\n",
    "    \"end\": df[\"n_end\"].astype(int),\n",
    "    \"score\": df[\"negLogQ\"],\n",
    "    \"strand\": \"+\",\n",
    "    \"frame\": \".\",\n",
    "    \"attribute\": \"ID=\" + df[\"Chr\"] + \";comments=\" + df[\"Chr\"] \n",
    "})\n",
    "\n",
    "# Creating the GFF formatted content\n",
    "gff_content = com_df.to_csv(sep='\\t', index=False, header=False)\n",
    "gff_file_path = './compare_pre_GFF.txt'\n",
    "with open(gff_file_path, 'w') as file:\n",
    "    file.write(gff_content)\n",
    "\n",
    "gff_file_path\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>q values</th>\n",
       "      <th>Description</th>\n",
       "      <th>negLogQ</th>\n",
       "      <th>Chromosome_num</th>\n",
       "      <th>type</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39109153</td>\n",
       "      <td>41523062</td>\n",
       "      <td>1.280000e-21</td>\n",
       "      <td>1p34.2(MYCL1)</td>\n",
       "      <td>20.892790</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148646436</td>\n",
       "      <td>149486905</td>\n",
       "      <td>5.190000e-18</td>\n",
       "      <td>1q21.2(MCL1)</td>\n",
       "      <td>17.284833</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>232526850</td>\n",
       "      <td>233617929</td>\n",
       "      <td>8.670000e-07</td>\n",
       "      <td>1q42.3(IRF2BP2)</td>\n",
       "      <td>6.061981</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154088696</td>\n",
       "      <td>154329131</td>\n",
       "      <td>3.880000e-09</td>\n",
       "      <td>---</td>\n",
       "      <td>8.411168</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36338557</td>\n",
       "      <td>36480225</td>\n",
       "      <td>2.120000e-11</td>\n",
       "      <td>1p34.3</td>\n",
       "      <td>10.673664</td>\n",
       "      <td>1</td>\n",
       "      <td>amp</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chr      Start        End      q values      Description    negLogQ  \\\n",
       "0  chr1   39109153   41523062  1.280000e-21    1p34.2(MYCL1)  20.892790   \n",
       "1  chr1  148646436  149486905  5.190000e-18     1q21.2(MCL1)  17.284833   \n",
       "2  chr1  232526850  233617929  8.670000e-07  1q42.3(IRF2BP2)   6.061981   \n",
       "3  chr1  154088696  154329131  3.880000e-09              ---   8.411168   \n",
       "4  chr1   36338557   36480225  2.120000e-11           1p34.3  10.673664   \n",
       "\n",
       "   Chromosome_num type Strand  \n",
       "0               1  amp      +  \n",
       "1               1  amp      +  \n",
       "2               1  amp      +  \n",
       "3               1  amp      +  \n",
       "4               1  amp      +  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"final_gene.csv\")\n",
    "df.drop(columns=df.columns[0], axis=1, inplace=True)\n",
    "# df.drop(columns=df.columns[4], axis=1, inplace=True)\n",
    "# df.drop(columns=df.columns[4], axis=1, inplace=True)\n",
    "# df.drop(columns=df.columns[5], axis=1, inplace=True)\n",
    "\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## now create a txt file that stores all the genes with descriptions. \n",
    "\n",
    "When making compare amp/del graph, we need to paste our txt data to a self-make gff file while adding the region information based on comparasion.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./compare_with_description_GFF.txt'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"final_gene.csv\")\n",
    "df.drop(columns=df.columns[0], axis=1, inplace=True)\n",
    "# df.drop(columns=df.columns[4], axis=1, inplace=True)\n",
    "# df.drop(columns=df.columns[5], axis=1, inplace=True)\n",
    "# df.drop(columns=df.columns[6], axis=1, inplace=True)\n",
    "\n",
    "df.head()\n",
    "\n",
    "df['dist'] = df['End'] - df['Start']\n",
    "\n",
    "# Determine the maximum distance for each chromosome and add it as a new column\n",
    "df['max_distance'] = df.groupby('Chr')['dist'].transform('max')\n",
    "\n",
    "# Add the minimum start value for each chromosome to the DataFrame\n",
    "df['min_start'] = df.groupby('Chr')['Start'].transform('min')\n",
    "\n",
    "# Add the maximum end value for each chromosome to the DataFrame\n",
    "df['max_end'] = df.groupby('Chr')['End'].transform('max')\n",
    "\n",
    "df['n_start'] = ((df['Start'] - df['min_start']) * df['max_distance']) / (df['max_end'] - df['min_start'])\n",
    "df['n_end'] = ((df['End'] - df['min_start']) * df['max_distance']) / (df['max_end'] - df['min_start'])\n",
    "\n",
    "\n",
    "com_df = pd.DataFrame({\n",
    "    \"seqname\": df[\"Chr\"],\n",
    "    \"source\": \"ovarian\",\n",
    "    \"feature\": df[\"type\"],\n",
    "    \"start\": df[\"n_start\"].astype(int).clip(lower=1),\n",
    "    \"end\": df[\"n_end\"].astype(int),\n",
    "    \"score\": df[\"negLogQ\"],\n",
    "    \"strand\": \"+\",\n",
    "    \"frame\": \".\",\n",
    "    \"attribute\": \"ID=\" + df[\"Chr\"] + \";comments=\" + df[\"Chr\"] + \"_\" + df[\"Description\"]\n",
    "})\n",
    "\n",
    "# Creating the GFF formatted content\n",
    "gff_content = com_df.to_csv(sep='\\t', index=False, header=False)\n",
    "gff_file_path = './compare_with_description_GFF.txt'\n",
    "with open(gff_file_path, 'w') as file:\n",
    "    file.write(gff_content)\n",
    "\n",
    "gff_file_path\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remeber to manually copy the txt file data into a self-made .gff file. And add the line of \n",
    "\n",
    "ovarian\tpython\tregion\t1\t60942\t.\t+\t.\tID=NC_000902.1:1..60942;Dbxref=taxon:97081;gbkey=Src;genome=proviral;mol_type=genomic DNA;nat-host=Escherichia coli O157:H7;synonym=Bacteriophage VT2-Sa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## But we still need the range for all the chr, so we need to find max and min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 247249719\n"
     ]
    }
   ],
   "source": [
    "print(result[\"Start\"].min(), result[\"End\"].max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "241440102"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr1 = result[result[\"Chr\"]==\"chr1\"]\n",
    "chr1[\"Start\"].max()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
